src/OpenLoco/GameCommands/RenameStation.cpp

```diff
@@ -0,0 +1,115 @@
+     */
+    static uint32_t renameStation(const uint8_t flags, StationId_t stationId, int16_t index, uint32_t buffer0, uint32_t buffer1, uint32_t buffer2)
+    {
+        GameCommands::setExpenditureType(ExpenditureType::Miscellaneous);
+
+        // Keep track of the station id over several calls.
+        static StationId_t _stationId{};
+        if (index == 1)
+            _stationId = stationId;
+
+        static uint32_t renameBuffer[9];
+
+        // Fill buffer over calls into the renameBuffer
+        if ((flags & GameCommands::Flags::apply) != 0)
+        {
+            static const std::array<int, 3> transformTable = { 2, 0, 1 };
+            int arrayIndex = transformTable.at(index);
+            renameBuffer[arrayIndex * 3] = buffer0;
+            renameBuffer[arrayIndex * 3 + 1] = buffer1;
+            renameBuffer[arrayIndex * 3 + 2] = buffer2;
+        }
+
+        // Applying the buffer?
+        if (index != 0)
+            return 0;
+
+        char renameStringBuffer[37] = "";
+        memcpy(renameStringBuffer, renameBuffer, sizeof(renameStringBuffer));
+        renameStringBuffer[36] = '\0';
+
+        // Figure out the current name for this station.
+        char currentStationName[256] = "";
+        auto station = StationManager::get(_stationId);
+        auto args = FormatArguments::common(station->town);
+        StringManager::formatString(currentStationName, station->name, &args);
+
+        // Verify the new name actually differs from the old one.
+        if (strcmp(currentStationName, renameStringBuffer) == 0)
+            return 0;
+
+        // If an empty string is given, generate one instead.
+        if (strlen(renameStringBuffer) == 0)
+        {
+            // Bailing out early?
+            if ((flags & GameCommands::Flags::apply) == 0)
+                return 0;
+
+            string_id oldStringId = station->name;
+            station->name = StationManager::generateNewStationName(_stationId, station->town, Map::Pos3(station->x, station->y, station->z), 0);
+            StringManager::emptyUserString(oldStringId);
+
+            station->updateLabel();
+            Gfx::invalidateScreen();
+            return 0;
+        }
+
+        // Allocate a string id for the new name.
+        string_id allocatedStringId = StringManager::userStringAllocate(renameStringBuffer, 0);
+        if (allocatedStringId == StringIds::empty)
+            return GameCommands::FAILURE;
+
+        // Bailing out early?
+        if ((flags & GameCommands::Flags::apply) == 0)
+        {
+            StringManager::emptyUserString(allocatedStringId);
+            return 0;
+        }
+
+        // Apply the new name to the station.
+        string_id oldStringId = station->name;
+        station->name = allocatedStringId;
+        StringManager::emptyUserString(oldStringId);
+
+        // Recalculate labels for the station and (surrounding) stations.
+        station->updateLabel();
+        StationManager::updateLabels();
```

Comment:
Why this additional updateLabels? I dont see it in the assembly and it will just end up calling the updateLabel on this station twice.

Reply:
Yeah, that looks like a mistake. It will suffice just update the label for the current station, indeed.